iocage_lib/ioc_upgrade.py

```diff
@@ -112,7 +113,17 @@ def upgrade_jail(self):
         try:
             tmp = tempfile.NamedTemporaryFile(delete=False)
             with urllib.request.urlopen(f) as fbsd_update:
-                tmp.write(fbsd_update.read())
+                # Since r345055 (11.3-RELEASE) freebsd-update
+                # returns 1 in case fetch needs to be called,
+                # this works around this problem.
+                exit_code_regex = \
+                    rb'''\t\t\techo "Run \'\$0 fetch\' first."\n'''\
+                    rb'''\t\t\texit 1'''
+                exit_code_sub = \
+                    rb'''\t\t\techo "Run '$0 fetch' first."\n'''\
+                    rb'''\t\t\t#exit 1 #iocage'''
```

Comment:
Hmm this is a tough problem, because if any output on their end changes, this all breaks. Parsing userland output is suboptimal, but I can't think of a better solution either. Ideally they don't continually change this script for other programs, without thinking about other consumers of it.

Reply:
Well, they changed it once in 2017 and now changed it back :). I assume this will be relatively stable (and they way this is implemented it won't break). Ideally we could add an environment variable to freebsd-update to control this behaviour (I might create a patch let's see), but this won't help versions in existence (like 11.3-RELEASE). I was a bit surprised iocage fetches it from github though and not from a direct freebsd resource (given that patches are fetched from there, it would make sense to not introduce an additional resource). Outside of the scope of this patch, but still.

Reply:
p.s. This is not parsing userland output, but modifying freebsd-update to not exit with this code. Parsing userland output would've been more unstable and in this case really hard (like parse it in python while still having full interactivity). And as this is basically just a replace, if the code changes in a compatible way, it will simply be a no-op.

Reply:
Something in that script we can call in the environment would be more ideal I think. As far as fetching from github, it was simply because I couldn't find a mirror of freebsd-update that would get the minor patches as the major lived. 

I'm open to changing that behavior of course! So with your suggestion I think we should do the following:
- Use an env variable or flag to freebsd-update to *not* exit 1 on these things
- Only run this sub on 11.3-RELEASE and under.

That should help us with stability in regards to this behavior as those shouldn't change at all.

Reply:
> p.s. This is not parsing userland output, but modifying freebsd-update to not exit with this code. Parsing userland output would've been more unstable and in this case really hard (like parse it in python while still having full interactivity). And as this is basically just a replace, if the code changes in a compatible way, it will simply be a no-op.

Yeah I noticed I made a mistake and this is actually the scripts content but I didn't want to edit it and make any reply look out of context :D

Reply:
I created a patch to freebsd-update and opened a bug:
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=239997

If approved I would modify this pull request to make use of it.

Reply:
Thanks for all the effort @grembo 👍

Reply:
@skarekrow After discussing the issue with @cperciva, I took his advise and implemented a different fix, that doesn't require patching freebsd-update on download and is backwards compatible. It looks for the symlink created by freebsd-update to determine if there is anything left to install (this is a stable enough interface to freebsd-update at this point). This also has the advantage of replacing the for loop hardcoded to three iterations with a while condition and also removing error messages from the log. I'll still look into patching freebsd-update to add a function called *freebsd-update updatesready*, but more as a new feature and not as a pre-condition to fix iocage. I feel like getting this in (also in the ports tree as a patch to 1.1) should be a priority, as it's actually breaking people's setup on upgrade to 11.3 at the moment.

Reply:
Agreed, this is something that has been a pain point for a lot of folks recently.